gem/impero_utils.py

```diff
@@ -316,3 +317,80 @@ def recurse_top(o):
         raise AssertionError("unhandled operation: %s" % type(op))
 
     return counter
+
+
+maxtype = partial(numpy.find_common_type, [])
+
+
+@singledispatch
+def _assign_dtype(expression, self):
+    return maxtype(map(self, expression.children))
+
+
+@_assign_dtype.register(gem.Terminal)
+def _assign_dtype_terminal(expression, self):
+    return self.scalar_type
+
+
+@_assign_dtype.register(gem.Zero)
+@_assign_dtype.register(gem.Identity)
+@_assign_dtype.register(gem.Delta)
+def _assign_dtype_real(expression, self):
+    return self.real_type
+
+
+@_assign_dtype.register(gem.Literal)
+def _assign_dtype_identity(expression, self):
+    return expression.array.dtype
+
+
+@_assign_dtype.register(gem.Power)
+def _assign_dtype_power(expression, self):
+    # Conservative
+    return self.scalar_type
+
+
+@_assign_dtype.register(gem.MathFunction)
+def _assign_dtype_mathfunction(expression, self):
+    if expression.name in {"real", "imag"}:
+        return self.real_type
+    else:
+        return maxtype(map(self, expression.children))
```

Comment:
Is this the safe thing to do? I don't think we support any mathfunctions that complexify real values in complex mode right now.

Reply:
What about square root of negatives?

Reply:
Oh yeah. Probably what I should do is handle the entire "math-table" one by one (which suggests migrating this mapper into the loopy codegen part [this is OK, since I didn't fix up the coffee backend]).

Reply:
Done.